asset_dashboard/templates/asset_dashboard/partials/forms/add_edit_phase_form.html

```diff
@@ -129,6 +154,239 @@ <h3>Phase Assets</h3>
+        hideError(sourceError)
+        hideError(securedError)
+        
+        if (year.toString().length != 4 || !year){
+          valid = false
+          yearError.innerHTML = 'Please enter a valid year'
+          showError(yearError)
+        } else {
+          hideError(yearError)
+        }
+
+        if (budget <= 0){
+          valid = false
+          budgetError.innerHTML = 'Please enter a valid amount'
+          showError(budgetError)
+        } else {
+          hideError(budgetError)
+        }
+
+        if (actual_cost <= 0){
+          valid = false
+          costError.innerHTML = 'Please enter a valid amount'
+          showError(costError)
+        } else {
+          hideError(costError)
+        }
+      }
+
+      // Back end validation and submission of forms
+      if (valid) {
+        let backendValid = true
+        let fetches = []
+
+        for (i = 0; i < fundings.length; i++) {
+          let funding = fundings[i]
+          const yearError = funding[0].nextElementSibling
+          const budgetError = funding[1].nextElementSibling.nextElementSibling
+          const costError = funding[3].nextElementSibling.nextElementSibling
+          const sourceError = funding[5].nextElementSibling
+          const securedError = funding[6].nextElementSibling
+
+          const errors = [yearError, budgetError, costError, sourceError, securedError]
+          errors.forEach(error => {
+            // Hide errors at the start of submission
+            error.classList.remove("visible")
+            error.setAttribute("aria-hidden","true")
+          })
+
+          let data = {
+            'year': funding[0].value,
+            'budget': funding[1].value,
+            'actual_cost': funding[3].value,
+            'source_type': funding[5].value,
+            'funding_secured': funding[6].value,
+            'phase': props.phase_id,
+            'id': funding.id ? funding.id : null,
+          }
+
+          // Get all fetches into an array, so we can wait to check
+          // if they were successful after they're all done
+          fetches.push(
+            fetch(`/projects/phases/fundingstream/`, {
+              method: 'POST',
+              body: JSON.stringify(data),
+              headers: {
+                'X-CSRFToken': getCookie('csrftoken'),
+                'Accept': 'application/json',
+                'Content-Type': 'application/json',
+              },
+              mode: 'same-origin' 
+            }).then((response)=> {
+              if (response.status != 201) {
+                backendValid = false
+              }
+              return response.json()
+            }).then(data => {
+              if (!backendValid) {
+                showBackendErrors(data, ...errors)
+              }
+            })
+          )
+        }
+
+        // Wait until all fetches run, then either show serverError or reload page
+        Promise.all(fetches).then(function() {
```

Comment:
this is a clever way to handle this — i like!

Reply:
thanks! i can see myself definitely coming back to this way of waiting for multiple fetches in the future haha